extensions/amp-form/0.1/form-validators.js

```diff
@@ -0,0 +1,370 @@
+
+  /** @override */
+  onInput(event) {
+    if (!this.validationBubble_.isActiveOn(event.target)) {
+      return;
+    }
+
+    if (event.target.checkValidity()) {
+      event.target.removeAttribute('aria-invalid');
+      this.validationBubble_.hide();
+    } else {
+      event.target.setAttribute('aria-invalid', 'true');
+      this.validationBubble_.show(event.target, event.target.validationMessage);
+    }
+  }
+}
+
+
+/**
+ * @abstract
+ */
+class AbstractCustomValidator extends FormValidator {
+
+  constructor(form) {
+    super(form);
+
+    /** @private @const {!Object<string, !Element>} */
+    this.inputValidationsDict_ = {};
+
+    /** @private @const {!Array<!Element>} */
+    this.inputVisibleValidationDict_ = {};
+  }
+
+  /**
+   * @param {!Element} input
+   */
+  reportInput(input) {
+    const invalidType = getInvalidType(input);
+    if (invalidType) {
+      this.showValidationFor(input, invalidType);
+    }
+  }
+
+  /**
+   * Hides all validation messages.
+   */
+  hideAllValidations() {
+    const inputs = this.form.querySelectorAll('input,select,textarea');
+    for (let i = 0; i < inputs.length; i++) {
+      this.hideValidationFor(inputs[i]);
+    }
+  }
+
+  /**
+   * @param {!Element} input
+   * @param {!string} invalidType
+   * @return {?Element}
+   */
+  getValidationFor(input, invalidType) {
+    if (!input.id) {
+      return null;
+    }
+    const selector = `[visible-when-invalid=${invalidType}]` +
+        `[validation-for=${input.id}]`;
+    if (!this.inputValidationsDict_[selector]) {
+      this.inputValidationsDict_[selector] = this.doc.querySelector(selector);
+    }
+    return this.inputValidationsDict_[selector];
+  }
+
+  /**
+   * @param {!Element} input
+   * @param {string} invalidType
+   */
+  showValidationFor(input, invalidType) {
+    const validation = this.getValidationFor(input, invalidType);
+    if (!validation) {
+      return;
+    }
+
+    if (!validation.textContent.trim()) {
+      validation.textContent = input.validationMessage;
+    }
+
+    input.setAttribute('aria-invalid', 'true');
+    validation.classList.add('visible');
+    this.inputVisibleValidationDict_[input.id] = validation;
+  }
+
+  /**
+   * @param {!Element} input
+   */
+  hideValidationFor(input) {
+    const visibleValidation = this.getVisibleValidationFor(input);
+    if (!visibleValidation) {
+      return;
+    }
+
+    input.removeAttribute('aria-invalid');
+    visibleValidation.classList.remove('visible');
+    delete this.inputVisibleValidationDict_[input.id];
+  }
+
+  /**
+   * @param {!Element} input
+   * @return {?Element}
+   */
+  getVisibleValidationFor(input) {
+    if (!input.id) {
+      return null;
+    }
+    return this.inputVisibleValidationDict_[input.id];
+  }
+
+  /**
+   * Whether an input should validate after an interaction.
+   * @param {!Element} input
+   * @return {boolean}
+   */
+  shouldValidateOnInteraction(input) {
+    return !!this.getVisibleValidationFor(input);
```

Comment:
Again this might be a misunderstanding of what I meant here. This basically about reporting the validation not actually doing the validation. And on interactions we don't report validations (similar to default browser behavior) only after the input has been reported (i.e. bubble or message is visible) we re-validate on consequent interactions.

So in this case this will re-report on every interaction ONLY if the input is currently being reported already. This is exactly the behavior of default reporting in browsers.